javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,230 @@
+
+```javascript
+function stringToNumber(string) {
+  let hashCode = 0;
+  for (let i = 0; i < string.length; i++) {
+    hashCode += string.charCodeAt(i);
+  }
+
+  return hashCode;
+}
+
+function hash(name, surname) {
+  return stringToNumber(name) + stringToNumber(surname);
+}
+```
+
+We not only consider the first letters with this technique. Instead, we take the entire name and convert it into numbers.
+
+You might be thinking, wouldn't it be just better to save the whole name as a hash code? That is true. This would make it unique for each name, but there are some good reasons to choose a number:
+
+- We do not save the hash code. The hash code is a calculation to find out which bucket (storage) our value will be placed in. Think about it as a locker number for example.
+
+- We save computation time. If we are to find where our hash is stored, it would be much faster and easier for a computer to compare two numbers rather than a string (or objects) to find the bucket where we have to store our elements (you will learn more about buckets shortly). Finding a bucket using a number would allow us to use that number as our index.
+
+- By using number as hash code, we can have hash codes that is uniform across variables where it doesn't have to only be a string, we can even hash numbers, or objects then we need an algorithm to generate a hash code number for our needs from an object, taking `class Person` object as an example, our hash code can be a combination of `strinngToNummber(name) + age`.
+
+### Buckets
+
+Buckets are storage that we need to store our elements. Simply, it's an array. From now on, we will be referring to our storage as buckets. We have a set of buckets as our storage, since our hash function now produce a number we are going to find out which bucket we gonna use for storage using that number as index. We receive a key `"Fred"` We hash it using our hash function, it produces the number `508` we look which Bucket has index `508` we put `"Fred"` in the bucket. This is the simple form, but there are more mechanics that we need to deal with.
+
+To get a value using a key, we put each entry inside a bucket as a `Node` item, which hold both the key and the value. To retrieve the value we hash a key, we find the bucket number, if the bucket is not empty then we go to that bucket we compare if the Node's key is the same key that is already in the bucket, if it is then we retrieve the Node's value otherwise we return null.
+
+Maybe you are wondering, why are we comparing the keys if we already found the hash code of that bucket? That is because remember, hash code is just the location, different items might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.
```

Comment:
You mean to replace the Bucket section entirely with a note? I believe it's too lengthy to be looked at as a note, unless we shorten it down. I don't think it hurts to keep it as a section, it fits nicely with all the bucket foreshadowing we have been doing in earlier sections.

Reply:
> Maybe you are wondering, why are we comparing the keys if we already found the hash code of that bucket? That is because remember, hash code is just the location, different items might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.

^ I just mean this